### PR TITLE
remove select query

### PR DESCRIPTION
**Pull request by clintropolis**
### Description

This PR removes the Select query type, in favor of the Scan query. Since the introduction of time ordered scan query with #7133, our documentation has already been recommending people use scan queries instead of select due to the heavy memory impact of the select query, and Druid SQL has not generated select queries at all since #7373. Functionally, scan is equivalent, less the paging functionality, however the more efficient nature of the scan query makes the paging functionality un-necessary.

Additionally, I feel that removing the choice between 2 _very_ similar query types should be more user friendly, and for code hygiene and lower maintenance for the query processing layer on our end.

This is obviously incompatible with previous releases so will need to be called out prominently in the release notes should we merge this PR.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.

### CODE REVIEW

**Comment by vogievetsky** [comment]
Very supportive of this in general!
I can not tell if you already do this from the diff but I think not: would it make sense to have a useful error message if someone does try to issue a select query explaining that it has been removed?

**Comment by fjy** [comment]
Closing/opening to restart teamcity

**Comment by clintropolis** [comment]
Not sure why it is failing teamcity consistently with the same errors in files I didn't modify, merged latest master in to see if it changes things.

**Comment by clintropolis** [comment]
>I can not tell if you already do this from the diff but I think not: would it make sense to have a useful error message if someone does try to issue a select query explaining that it has been removed?

I did not add any special handling for this, so it will show the default error:

```
Could not resolve type id 'select' into a subtype of [simple type, class org.apache.druid.query.Query]: known type ids = [Query, dataSourceMetadata, groupBy, scan, search, segmentMetadata, timeBoundary, timeseries, topN] at [Source: HttpInputOverHTTP@60001e07[c=336,q=0,[0]=null,s=STREAM]; line: 1, column: 2]
```

I guess I could add a `RemovedQuery` class or something to that effect, with an implementation explodes into a more specific error about a query type previously existing but is obsolete and has been removed, and bind it to `select` and anything else we ever remove. However, that also means I have to add some code and tests, and that makes me sad since I was just trying to nuke a bunch of code in this PR... but will add it if we really think it is necessary.

**Comment by clintropolis** [comment]
Ah, the teamcity failures are actually legitimate. The two errors are methods that only the select query engine was using: a variant of `DefaultDimensionSpec.toSpec` and `Cursor.advanceTo`.

**Comment by clintropolis** [comment]
>I can not tell if you already do this from the diff but I think not: would it make sense to have a useful error message if someone does try to issue a select query explaining that it has been removed?

Ok, added back a `SelectQuery` class that implements `Query` and only knows how to make `RuntimeException` and a test to make sure that it happens on deserialization. Trying to use a select query will result in an error message of this form:

```
com.fasterxml.jackson.databind.JsonMappingException: Instantiation of [simple type, class org.apache.druid.query.select.SelectQuery] value failed: The 'select' query has been removed, use 'scan' instead. See https://druid.apache.org/docs/latest/querying/select-query.html for more details.
```

All it cost was adding 235 lines of code 😞

File: docs/configuration/index.md [hunk 1]
```diff
@@ -1544,7 +1544,7 @@ You can optionally only configure caching to be enabled on the Broker by setting
 |`druid.broker.cache.useResultLevelCache`|true, false|Enable result level caching on the Broker.|false|
 |`druid.broker.cache.populateResultLevelCache`|true, false|Populate the result level cache on the Broker.|false|
 |`druid.broker.cache.resultLevelCacheLimit`|positive integer|Maximum size of query response that can be cached.|`Integer.MAX_VALUE`|
-|`druid.broker.cache.unCacheable`|All druid query types|All query types to not cache.|`["groupBy", "select"]`|
```
Comment (gianm) [inline docs/configuration/index.md]: Nice.

File: docs/querying/select-query.md [hunk 2]
```diff
@@ -24,235 +24,11 @@ sidebar_label: "Select"
   -->
 
 
-> We encourage you to use the [Scan query](../querying/scan-query.md) type rather than Select whenever possible.
-> In situations involving larger numbers of segments, the Select query can have very high memory and performance overhead.
+> The native Druid Select query has been removed in favor of the [Scan query](../querying/scan-query.md), which should
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply docs/querying/select-query.md]: Updated with your suggestion :+1:

**Comment by clintropolis** [comment]
>I'm wondering whether we really need this. Should we remove SelectQuery at some point anyway because we cannot keep all legacy classes forever? A nice error message is always good for Druid users, but I guess this may not be needed since we will call out in the release notes and keep the doc for removed select query.

I am in this camp as well, originally the PR just removed it all, but I guess this is a nicer user experience. I do think we could remove this placeholder in a release or two after the next one.

File: processing/src/main/java/org/apache/druid/query/select/SelectQuery.java [hunk 3]
```diff
@@ -20,212 +20,150 @@
...
 
 import java.util.List;
 import java.util.Map;
-import java.util.Objects;
 
-/**
- */
-@JsonTypeName("select")
-public class SelectQuery extends BaseQuery<Result<SelectResultValue>>
+public class SelectQuery implements Query<Object>
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/select/SelectQuery.java]: added

**Comment by gianm** [comment]
> I'm wondering whether we really need this. Should we remove `SelectQuery` at some point anyway because we cannot keep all legacy classes forever? A nice error message is always good for Druid users, but I guess this may not be needed since we will call out in the release notes and keep the doc for removed select query.

I would suggest keeping the error message for at least a few releases. The Select query, even though flawed, was somewhat popular and I think this message will reduce confusion in the user base.

**Pull request merged**